src/support/scratch.c

```diff
@@ -40,14 +40,8 @@ __wt_buf_grow_worker(WT_SESSION_IMPL *session, WT_ITEM *buf, size_t size)
 	 * This function is also used to ensure data is local to the buffer,
 	 * check to see if we actually need to grow anything.
 	 */
-	if (size > buf->memsize) {
-		if (F_ISSET(buf, WT_ITEM_ALIGNED))
-			WT_RET(__wt_realloc_aligned(
-			    session, &buf->memsize, size, &buf->mem));
-		else
-			WT_RET(__wt_realloc(
-			    session, &buf->memsize, size, &buf->mem));
-	}
+	if (size > buf->memsize)
+		WT_RET(__wt_realloc_item(session, buf, size));
```

Comment:
> What's the motivation for this change?

To add a `__wt_malloc` function so we can iterate forward on not clearing memory that doesn't need to be cleared.

I also changed the `__wt_strndup` function to use `__wt_malloc` (because it was easy), and changed read/write operations to not clear their buffers (because I think that's likely to have the largest impact on performance).

The latter change wasn't as simple as I'd like -- the simplest solution I found was based on the facts that I/O is always done into `WT_ITEM` structures, and I didn't see any place we ever need to clear `WT_ITEM` structure memory. So, I pushed a `WT_ITEM` reallocation API down into the system-dependent allocation code and that path optionally aligns buffers and never clears memory. That means we won't clear scratch buffers or I/O buffers, which seems about right.

> We set the WT_ITEM_ALIGNED flag to indicate whether alignment is required: this is now ignoring that flag.

This change moves the `WT_ITEM_ALIGNED` flag test  to the `__wt_realloc_item` function.

> Can we do a better job of only aligning allocations that could be used for read/write operations?

I didn't investigate that, I left our current approach unchanged.